src/app/components/date-range-picker/date-range-picker.component.ts

```diff
@@ -0,0 +1,143 @@
+	ValidationErrors,
+	Validators,
+} from '@angular/forms';
+import { Subject, takeUntil, tap } from 'rxjs';
+import { DateRange } from '../../models';
+import { ErrorHandlerService } from '../../services';
+
+@Component({
+	selector: 'app-date-range-picker',
+	templateUrl: './date-range-picker.component.html',
+	styleUrls: ['./date-range-picker.component.scss'],
+	changeDetection: ChangeDetectionStrategy.OnPush,
+	providers: [
+		{
+			provide: NG_VALIDATORS,
+			useExisting: forwardRef(() => DateRangePickerComponent),
+			multi: true,
+		},
+		{
+			provide: NG_VALUE_ACCESSOR,
+			useExisting: forwardRef(() => DateRangePickerComponent),
+			multi: true,
+		},
+	],
+})
+export class DateRangePickerComponent implements ControlValueAccessor, OnDestroy {
+	private readonly errorHandlerService = inject(ErrorHandlerService);
+
+	// inner FormGroup used to manage ControlValueAccessor value
+	readonly dateRangeFormGroup = this.getDateRangeFormGroup();
+
+	readonly minDate = new Date(2015, 0, 1);
+	readonly maxDate = new Date();
+
+	// Start and End ErrorHandlers used to populate `mat-error`
+	readonly startDateErrorsHandler = this.errorHandlerService.getDatepickerErrorsHandler('Start Date');
+	readonly endDateErrorsHandler = this.errorHandlerService.getDatepickerErrorsHandler('End Date');
+
+	// onChange callback that will be overriden using `registerOnChange`
+	onChange: (dateRange: DateRange) => void = () => {
+		/** empty */
+	};
+	// onTouched callback that will be overriden using `registerOnTouched`
+	onTouched = () => {
+		/** empty */
+	};
+
+	/**
+	 * Bubbles up MatDateRangePicker errors to this
+	 * Component's ControlValueAccessor validators
+	 */
+	matDateRangePickerValidators = (): ValidationErrors | null => {
+		const nestedErrors =
+			this.dateRangeFormGroup.errors ??
+			this.dateRangeFormGroup.controls.start.errors ??
+			this.dateRangeFormGroup.controls.end.errors;
+
+		if (nestedErrors) {
+			return nestedErrors;
+		}
+
+		return null;
+	};
+
+	private readonly destroyer$ = new Subject<void>();
+
+	constructor() {
+		this.dateRangeFormGroup.valueChanges
+			.pipe(
+				tap(({ start, end }) => {
+					if (!this.dateRangeFormGroup.valid) {
+						return;
+					}
+
+					if (!start || !end) {
+						return;
+					}
+
+					this.onChange([start, end]);
+				}),
+				takeUntil(this.destroyer$)
+			)
+			.subscribe();
+	}
+
+	/**
+	 * Creates FormGroup that is responsible for setting
+	 * Component's ControlValueAccessor value
+	 */
+	getDateRangeFormGroup(): FormGroup<{
+		start: FormControl<Date | null>;
+		end: FormControl<Date | null>;
+	}> {
+		return new FormGroup({
+			start: new FormControl<Date | null>(null, {
+				validators: Validators.required,
+			}),
+			end: new FormControl<Date | null>(null, {
+				validators: Validators.required,
+			}),
+		});
+	}
+
+	/**
+	 * Set Component's ControlValueAccessor value
+	 */
+	writeValue([start, end]: DateRange): void {
+		this.dateRangeFormGroup.setValue({ start, end });
+	}
```

Comment:
This is not necessarily an error, I just want to call out, that if the Datepicker component were reused in multiple places, and the parent component would add a `null` initial value instead of `this.getInitialDateRange()` to the form control, then at line 114 you would actually try to destruct a null/undefined value that would lead to an error.
So as a fix it may be `writeValue([start, end]?: DateRange):` -> using the optional `?` operator

Reply:
I think this is a fair consideration, given that the default behavior is to set a control's value to null.

I wonder how fair down the rabbit hole should go, would it be fair enough to support nullish and [Date, Date]? And throw error if an unexpected value is used instead?

Reply:
I've started a stackblitz testing how `@angular/material` deals with other unexpected values for their single date picker:
https://stackblitz.com/edit/angular-5ab9cb?file=src%2Fapp%2Fdatepicker-overview-example.html,src%2Fapp%2Fdatepicker-overview-example.ts

Looks like whenever an "invalid" non-date value is used, the ui doesn't actually update

Reply:
My suggestion is if you have `[start, end]?: DateRange`, so that the value can be null, then you just set the `this.dateRangeFormGroup.setValue(null);`. It is not the problem, the problem is in the case when `[start, end]: DateRange`  will be null and if you will try to set `({ start, end }) ` (line 115) then I believe it will throw an error.

Reply:
Setting it to null caused a problem which is now handled. [null, null] is already covered